### PR TITLE
concatenate the one range with another range

### PR DESCRIPTION
**Pull request by bikramkawan**
@thinkh @sgratzl

### CODE REVIEW

**Comment by mstreit** [comment]
@bikramkawan, you can assign them as reviewers, see button on the right.

File: src/range/CompositeRange1D.ts [hunk 1]
```diff
@@ -36,6 +36,12 @@ export default class CompositeRange1D extends Range1D implements ICompositeRange
     return new CompositeRange1D(this.name, this.groups.map((g) => <Range1DGroup>g.union(other, size)), r);
   }
 
+  concat(other: Range1D, size?: number) {
+    const r = this.groups.length > 1 ? super.union(other, size) : undefined;
```
Comment (sgratzl) [inline src/range/CompositeRange1D.ts]: why is `super.union` called instead of `.concat`?

File: src/range/CompositeRange1D.ts [hunk 2]
```diff
@@ -36,6 +36,12 @@ export default class CompositeRange1D extends Range1D implements ICompositeRange
...
+    return new CompositeRange1D(this.name, this.groups.map((g) => <Range1DGroup>g.union(other, size)), r);
```
Comment (sgratzl) [inline src/range/CompositeRange1D.ts]: again, `union` instead of `concat`

**Comment by thinkh** [comment]
@mstreit He can't, because he isn't a collaborator to this repository.

File: src/range/Range1D.ts [hunk 3]
```diff
@@ -204,28 +204,59 @@ export default class Range1D {
...
+    return other.fromLike(r);
   }
 
   /**
+   * concatenate this range (first) with another range (second) without sorting
+   * @param other
+   * @param size
+   * @returns {Range1D}
+   */
+  concat(other: Range1D, size?: number): Range1D {
```
Comment (sgratzl) [inline src/range/Range1D.ts]: for me `concat` doesn't imply for me that there are no duplicates allowed. At least when I concatenate two strings, duplicates are allowed.

File: src/range/Range1DGroup.ts [hunk 4]
```diff
@@ -19,6 +19,11 @@ export default class Range1DGroup extends Range1D {
     return new Range1DGroup(this.name, this.color, r);
   }
 
+   concat(other: Range1D, size?: number): Range1DGroup {
+    const r = super.union(other, size);
```
Comment (sgratzl) [inline src/range/Range1DGroup.ts]: why is `super.union` called?

**Review by thinkh — DISMISSED** [review]

**Review by sgratzl — CHANGES_REQUESTED** [review]
several copy paste errors that needs to be fixed + I don't know whether `concat` is the right name for the logic that it implements

File: src/range/Range1D.ts (hunk 3 above)
Comment (thinkh) [reply src/range/Range1D.ts]: Would `join` be a better name?

Comment (sgratzl) [reply src/range/Range1D.ts]: there is already a join method: https://github.com/phovea/phovea_core/blob/develop/src/range/index.ts#L50

**Comment by sgratzl** [comment]
I don't get the motivation behind this method?

File: src/range/Range1D.ts (hunk 3 above)
Comment (thinkh) [reply src/range/Range1D.ts]: We tried the `join` function that you send. However, it joins multiple dimensions to a single. We need to join multiple Range1D to a single one.

**Comment by thinkh** [comment]
@sgratzl `Range1D.union()` is sorting the results, the proposed `Range1D.concat()` isn't. It's important for mothertable to *not* sort the results.

**Comment by sgratzl** [comment]
I get the functionality but not the usecase

**Comment by bikramkawan** [comment]
The use case is similar to this.
https://github.com/Caleydo/mothertable/issues/103
For example: 
The data is sorted by country > career status > gender. 
Consider we have already sorted country and career status which gives use array of Range.
What I was looking for is similar as below:
image
The index of each sorted is [0],[9] and so on. This means these are the array of Range Object. In the end we need to join to form single Range object which is exactly shown in last array [0,9.. 4,7]. 
The final  range is used to render the visualization.

**Comment by sgratzl** [comment]
why is the criteria "no-duplicates" that the union/concat method fulfills needed?

**Comment by bikramkawan** [comment]
what do you mean by no-duplicates here?

**Comment by sgratzl** [comment]
> @sgratzl Range1D.union() is sorting the results, the proposed Range1D.concat() isn't. 

union is a set operation, i.e. it ensures that there are no duplicates in the set and since it is a set, the order doesn't matter. 

the current version of `concat` uses 90% of the union logic also ensuring that there are no duplicates.

**Comment by bikramkawan** [comment]
Yes. it uses the logic of union but we need the union without the sorting. The current union method returns the indexes after sorting.

File: src/range/Range.ts [hunk 5]
```diff
@@ -94,6 +94,22 @@ export default class Range {
     return r;
   }
 
+
+  concat(other: Range, size?: number[]) {
+    if (this.isAll || other.isNone) {
```
Comment (sgratzl) [inline src/range/Range.ts:99]: does it make sense, if this or the other contain all ids, not to concatenate them?

File: src/range/Range1D.ts [hunk 6]
```diff
@@ -226,6 +228,28 @@ export default class Range1D {
...
+   */
+
+  concat(other: Range1D, size?: number): Range1D {
+    if (this.isAll || other.isNone) {
+      return this.clone();
+    }
+    if (other.isAll || this.isNone) {
+      return other.clone();
+    }
+    const r = this.iter(size).asList();
```
Comment (sgratzl) [inline src/range/Range1D.ts:244]: why need both list converted to their indices before combining them?

File: src/range/CompositeRange1D.ts [hunk 7]
```diff
@@ -36,6 +36,11 @@ export default class CompositeRange1D extends Range1D implements ICompositeRange
     return new CompositeRange1D(this.name, this.groups.map((g) => <Range1DGroup>g.union(other, size)), r);
   }
 
+  concat(other: Range1D, size?: number) {
+    const r = this.groups.length > 1 ? super.concat(other, size) : undefined;
+    return new CompositeRange1D(this.name, this.groups.map((g) => <Range1DGroup>g.concat(other, size)), r);
```
Comment (sgratzl) [inline src/range/CompositeRange1D.ts:41]: does it make sense to add the items to each group within a composite?

**Comment by thinkh** [comment]
@bikramkawan What is the status here? Do you still need that functionality for mothertable?

**Comment by bikramkawan** [comment]
@thinkh  Currently, I am using push list method. we can keep this for future reference if there is no any harm in our branch.

**Pull request closed without merging**